lib/streamlit/web/server/routes.py

```diff
@@ -64,33 +55,123 @@ def set_extra_headers(self, path: str) -> None:
+                self.set_status(416)  # Range Not Satisfiable
+                self.set_header("Content-Type", "text/plain")
+                self.set_header("Content-Range", "bytes */%s" % (size,))
+                return
+            if end is not None and end > size:
+                # Clients sometimes blindly use a large range to limit their
+                # download size; cap the endpoint at the actual file size.
+                end = size
+            # Note: only return HTTP 206 if less than the entire range has been
+            # requested. Not only is this semantically correct, but Chrome
+            # refuses to play audio if it gets an HTTP 206 in response to
+            # ``Range: bytes=0-``.
+            if size != (end or size) - (start or 0):
+                self.set_status(206)  # Partial Content
+                self.set_header(
+                    "Content-Range", httputil._get_content_range(start, end, size)
+                )
+        else:
+            start = end = None
+
+        if start is not None and end is not None:
+            content_length = end - start
+        elif end is not None:
+            content_length = end
+        elif start is not None:
+            content_length = size - start
+        else:
+            content_length = size
 
-            raise e
+        ################## BEGIN OVERRIDDEN SECTION OF TORNADO CODE ##################
 
-    def write_error(self, status_code: int, **kwargs) -> None:
-        if status_code == 404:
-            index_file = os.path.join(file_util.get_static_dir(), "index.html")
-            self.render(index_file)
+        if (
+            include_body
+            and self.default_filename
+            and self.absolute_path.endswith(self.default_filename)
+        ):
+            self.render_index_template()
+        else:
+            self.set_header("Content-Length", content_length)
+            if include_body:
+                content = self.get_content(self.absolute_path, start, end)
+                if isinstance(content, bytes):  # type:ignore[unreachable]
+                    content = [content]  # type:ignore[unreachable]
+                for chunk in content:
+                    try:
+                        self.write(chunk)
+                        await self.flush()
+                    except iostream.StreamClosedError:
+                        return
+            else:
+                assert self.request.method == "HEAD"
+        ################## END OVERRIDDEN SECTION OF TORNADO CODE ##################
+
+    def render_index_template(self):
+        """This method renders the Tornado template when static file is not found or the request is for index.html."""
+        base_path = config.get_option("server.baseUrlPath").strip("/")
+        if base_path:
+            base_path = f"/{base_path}/"
         else:
-            super().write_error(status_code, **kwargs)
+            base_path = "/"
```

Comment:
@lukasmasuch raised the valid concern yesterday that this might be a behavioral change and that we could break existing deployments by changing this (I think by default the value was `""` which is equal to not having a base tag). I guess we should try and test on different platforms what happens when we set `"/"` in a base tag

Reply:
Hmmmm...we should probably discuss this, because it is correct that we are effectively saying the `base` is the root rather than relative. The issue becomes now if it's relative, nested directory url paths just won't work, so in part, this behavioral change seems necessary to have everything work.

Now if people are proxying and not setting the `server.baseUrlPath` and doing weird stuff, this would break, but I cannot imagine this was anticipated behavior we might break, and the solution would be to set the `server.baseUrlPath` and have the proxy include that base.